packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```diff
@@ -762,6 +772,217 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
+      const { type, id } = object;
+      if (!this._allowedTypes.includes(type)) {
+        return {
+          tag: 'Left',
+          value: {
+            id,
+            type,
+            error: errorContent(SavedObjectsErrorHelpers.createUnsupportedTypeError(type)),
+          },
+        };
+      }
+      const requiresNamespacesCheck = this._registry.isMultiNamespace(type);
+      return {
+        tag: 'Right',
+        value: {
+          type,
+          id,
+          ...(requiresNamespacesCheck && { esRequestIndex: bulkGetRequestIndexCounter++ }),
+        },
+      };
+    });
+
+    // get multi-namespace saved objects.
+    const bulkGetDocs = expectedBulkGetResults
+      .filter(isRight)
+      .filter(({ value }) => value.esRequestIndex !== undefined)
+      .map(({ value: { type, id } }) => ({
+        _id: this._serializer.generateRawId(namespace, type, id),
+        _index: this.getIndexForType(type),
+        _source: ['type', 'namespaces'],
+      }));
+
+    const bulkGetResponse = bulkGetDocs.length
+      ? await this.client.mget({ body: { docs: bulkGetDocs } }, { ignore: [404], meta: true })
+      : undefined;
+    // fail fast if we can't verify a 404 response is from Elasticsearch
+    if (
+      bulkGetResponse &&
+      isNotFoundFromUnsupportedServer({
+        statusCode: bulkGetResponse.statusCode,
+        headers: bulkGetResponse.headers,
+      })
+    ) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundEsUnavailableError();
+    }
+
+    let bulkDeleteRequestIndexCounter = 0;
+    const bulkDeleteParams: BulkDeleteParams[] = [];
+
+    const expectedBulkDeleteResults = await Promise.all(
+      expectedBulkGetResults.map<Promise<ExpectedBulkDeleteResult>>(
+        async (expectedBulkGetResult) => {
+          if (isLeft(expectedBulkGetResult)) {
+            return { ...expectedBulkGetResult };
+          }
+          const { esRequestIndex, id, type, version } = expectedBulkGetResult.value;
+
+          let namespaces;
+          let versionProperties;
+
+          if (esRequestIndex !== undefined) {
+            const indexFound = bulkGetResponse?.statusCode !== 404;
+
+            const actualResult = indexFound
+              ? bulkGetResponse?.body.docs[esRequestIndex]
+              : undefined;
+
+            const docFound = indexFound && isMgetDoc(actualResult) && actualResult.found;
+
+            if (!docFound) {
+              return {
+                tag: 'Left',
+                value: {
+                  id,
+                  type,
+                  error: errorContent(
+                    SavedObjectsErrorHelpers.createGenericNotFoundError(type, id)
+                  ),
+                },
+              };
+            }
+            if (
+              // @ts-expect-error MultiGetHit is incorrectly missing _id, _source
+              !this.rawDocExistsInNamespace(actualResult, namespace) &&
+              !force
+            ) {
```

Comment:
Is that the correct check? The `force` option is meant to allow to delete objects that are in multiple namespaces, but still **including the namespace we're trying to delete from** 

If the object is not in the current namespace at all, we want to throw, like we do for the plain `delete` API:

https://github.com/elastic/kibana/blob/34c228b0c941308e01298e31223e7af0a5a2cee9/packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts#L693-L707

I think the check is wrong here, given you're checking if the object does not belong in the current NS, instead of checking if the object does not belong **exclusively** in the current NS?

Reply:
I interpreted the `force` behavior from how these checks are used in other SO repo methods and wasn't too sure about how to handle the flag in `bulkDelete`. 
Your description makes a lot more sense. Thanks!

Reply:
Hopefully, the condition is correct now.
Please ignore the new variable 'descriptiveNamesToFollowWhatTheCodeIsDoing'. I fully intend to rename them again during the refactor.